### DISCUSSION TITLE
Reregister Wizard.Scenes by command

### DISCUSSION THREAD

**Discussion by Ludenser** [telegraf/telegraf#1682] [Q&A]
In my case i have to update condition of scenes by command "/update" without restart the bot.
Now, ive got this:
```
module.exports = (ctx) => {

  const totalSceneInitComposer = new Composer();
  const stage = new Scenes.Stage([initialScene(ctx), ...complexScene(ctx)])

  totalSceneInitComposer.use(stage.middleware())
  return totalSceneInitComposer
}
```
```
composer.command('update', async (ctx) => {

  ctx.session.all_lists = []
  ctx.session.isAlreadyFilled = false
  if (ctx.session.isAuthUser) {
    await addTasksToCtx(ctx) 
    if (ctx.session.all_lists.length) {
      composer.use(totalSceneInitComposer(ctx))
    }
    await ctx.deleteMessage()
  } else {
    await sendProses(ctx, 'Нет доступа')
  }
})
```
But this doesnt work as i expect, and i have to reboot the bot, because the new ones is in conflict with the old ones scene's middlewares. Is there a possibility to drop the old ones scenes from bot, maybe? 
Thanx!

**Comment by MKRhere** [comment]
You shouldn't add new from within another middleware; that said, what is the real problem you're trying to solve? Maybe we can suggest an alternate approach.

**Reply by Ludenser** [reply]
Thx for answering, i appreciate this! 
Each scene is almost the same, but with different data inside. Scenes are "dynamically" registered based on data from the API. There are cases when the actual data on the server may change, and then you need to restart the entire bot. I'm looking for a way to avoid this. Now new scenes seem to exist, are being added, but there is no data in them.

**Reply by MKRhere** [reply]
Why not pass the data around using `ctx.session` and change behaviour without remounting the scenes?

**Reply by Ludenser** [reply]
Cause data already in ctx.session and the number of scenes is always variable, depending on this data. Number of scenes = ctx.session.data.length. Each scene which i have to update looks like this.
```
/**
 * @param {[Object]} arr
 * @param {[Object]} list
 */

module.exports = (arr, list) => {
  const complexSceneArray = _(arr).map((task, i) => {
    const complex_scene = new Composer();

    complex_scene.action('enter', async (ctx) => {
      const ClickAPI = new Clickup(ctx.session.user.CU_Token);

      try {
        await ctx.deleteMessage();
        await ClickAPI.Tasks.setStatus(task.id, 'in progress');
        await ClickAPI.TimeTracking.startEntry(task.id);
        await setAssigneeFeature(ctx.session.userName, task.id, ctx.session.user.CU_Token);
        await sendMessageRouteEnter(ctx, task.name, task.id);
        await getAttentionFeature(ctx, task.custom_fields);
      } catch (e) {
        await sendError(ctx, e);
      }
    });

    complex_scene.action('reenter', async (ctx) => {
      try {
        await ctx.deleteMessage();
        await sendMessageRouteEnter(ctx, task.name, task.id);
      } catch (e) {
        await sendError(ctx, e);
      }
    });

complex_scene.action('next_step', async (ctx) => {
      const ClickAPI = new Clickup(ctx.session.user.CU_Token);

      try {
        await ClickAPI.Tasks.setStatus(task.id, 'done');
        await ClickAPI.TimeTracking.stopEntry(task.id);
        await resolveAllCheckListsAndItems(task.checklists, 'true', ctx.session.user.CU_Token);
        await ClickAPI.TimeTracking.startEntry(list.mainTask[0].id);

        await ctx.deleteMessage();
        await sendMessageNextStep(ctx, task.name, arr[i + 1].name);
        await ctx.wizard.next();
      } catch (e) {
        await sendError(ctx, e);
      }
    });
return complex_scene;
  });

  return complexSceneArray;
};
```
This is not a complete example, there is still a lot of logic on handlers, but it is enough to understand.

then creating Scenes.WizardScene from this array of composers
```
const complexScene = (ctx) => {

  const wizardScene = _(ctx.session.all_lists)
    .map((list, i) => {
      
      return new Scenes.WizardScene(`ROUTE_${i}_WIZARD_ID`, ...complex_scene(ctx.session.all_lists[i].tasksWithoutMain, list, ctx.session))
    })

  return wizardScene
}

module.exports = complexScene
```
When data was changed i have to update this scenes without bot restart, because other users use bot at this moment and theyll have to interrupt  and restart the work((
This update usually must will one time per day and now i use sheduler to restart the bot during non-working hours, but sometimes I extremely need to do it manually without restart
